python/frontend/cirrus/cirrus/CrossValidation.py

```diff
@@ -0,0 +1,168 @@
+                       hyper_vars = hyper_vars,
+                       hyper_params = hyper_params,
+                       machines = machines,
+                       base_port = base_port,
+                       config_num = i)
+      self.cirrus_objs.append(task)
+      combos = len(list(itertools.product(*hyper_params)))
+      base_port += 2 * combos
+      self.infos.extend([{'color': get_random_color()} for i in range(2 * combos)])
+      task_command_dicts.append(task.get_command_dict())
+    for machine in self.machines:
+        for c_dict in task_command_dicts:
+            self.command_dict[machine[0]].extend(c_dict[machine[0]])
+    self.loss_lst = [[] for _ in range(self.get_number_experiments())]
+    self.num_threads = len(self.loss_lst)
+    print(self.machines)
+  
+  def run(self, UI=False):
+    command_dict_to_file(self.command_dict)
+    self.start_threads()
+    for c in self.cirrus_objs:
+        c.run(run=False)
+    if UI:
+      def ui_func(self):
+        graph.bundle = self
+        graph.app.run_server()
+      self.ui_thread = threading.Thread(target=ui_func, args = (self, ))
+      self.ui_thread.start()
+
+  def get_number_experiments(self):
+    return self.cirrus_objs[0].get_number_experiments()
+  
+  def get_number_experiments_ups(self):
+    return self.cirrus_objs[0].get_number_experiments() * self.k
+
+  def get_number_experiments_cps(self):
+    return self.cirrus_objs[0].get_number_experiments() * self.k
+
+  def get_info(self, i, param = None):
+    out = self.infos[i]
+    if param:
+      return out[param]
+    else:
+      return out
+
+  def start_threads(self):
+    def custodian(cirrus_objs, i):
+      while True:
+        total_ttl = {}
+        for cirrus_obj in cirrus_objs:
+          loss_lst = cirrus_obj.loss_lst[i]
+          for time_to_loss in loss_lst:
+            rounded_time = int(round(time_to_loss[0]))
+            if rounded_time in total_ttl:
+              total_ttl[rounded_time].append(time_to_loss[1])
+            else:
+              total_ttl[rounded_time] = [time_to_loss[1]]
+        avg_ttl = [(time, sum(total_ttl[time]) / len(total_ttl[time])) for time in total_ttl]
+        self.loss_lst[i] = sorted(avg_ttl, key = itemgetter(0))
+
+    for i in range(self.num_threads):
+      p = threading.Thread(target=custodian, args=(self.cirrus_objs, i))
+      p.start()
+    
+    
+    copy_threads = len(self.machines)
+    def copy_and_run(thread_id):
+        while True:
+            if thread_id >= len(self.machines):
+                return
+            sh_file = "machine_%d.sh" % thread_id
+            ubuntu_machine = "ubuntu@%s" % self.machines[thread_id][0]
+            print(thread_id, ubuntu_machine)
+
+            cmd = "scp %s %s:~/python_files" % (sh_file, ubuntu_machine)
+            print cmd
+            os.system(cmd)
+            cmd = 'ssh %s "chmod +x python_files/%s; ./python_files/%s &"' % (ubuntu_machine, sh_file, sh_file)
```

Comment:
scp and ssh commands should be run through a library (something like paramiko). Check automate.py